Asap7 update and flow fixes without `.lib`

joennlae opened the pull request:
To make it reviewable.

GIthub cannot render changes due to the swapped library files :-)

I worked heavily with ASAP7 for a project and thought I would upstream my changes.
I will try to reason why every commit makes sense but I am happy to discuss and also eager to see the results of the CI. We probably have to update things after CI again.

- floorplan: add `set_dont_use` before `repair_design`

If one uses `RESYNTH_TIMING_RECOVER` it used cells that we do not want to use during `repair_design` and introduced them to the design.

- cts: move `set_dont_use` before `repair_clock_inverters`

Same issue as before `repair_clock_inverters` cloud introduce cells we dont want to use.

- asap7: update to `asap7` `v28` + adding super low threshold option

Updating `.lib`, `.lef`, `.gds` for asap7 and `config.mk` to newest version.
GIthub cannot render this commit.

Should fix all these issues:

> ### There are some fixes in the newer version:
> 1. ICG lib with "clock gating function and state table" has been fixed. [ICG cells throw an error in library compiler The-OpenROAD-Project/asap7#3](https://github.com/The-OpenROAD-Project/asap7/issues/3)
>    Edit: ICG*DC* cells with incorrect width has been fixed. [ICG*DC* cells have incorrect width The-OpenROAD-Project/asap7#12](https://github.com/The-OpenROAD-Project/asap7/issues/12)
> 2. Large ASNYC FF delay has been fixed. [Wrong timing values (1e31) for ASYNC_DFFHx1_ASAP7_75t_R in the sc7p5t_SEQ_RVT_SS_nldm_201020 library The-OpenROAD-Project/asap7#9](https://github.com/The-OpenROAD-Project/asap7/issues/9)
> 3. Multi-VT GDS files missing has been fixed.

- global_route: remove deprecated option `-report_violating_nets`

- asap7: remove unused `yosys` blackbox files

Cleanup of the `.blackbox.v` files and references for `asap7` as we use `.lib` files

- asap7: allow with `v28` working clock gating cells `ICG*`

Allow ICG* cells as they have been fixed for version 28.

- asap7: make `CTS_BUF_DISTANCE` editable via parent config

Just to allow for more flexibility same as in other techs.

Now we come to the more opinionated commits. Which I am also very happy to discuss.

- global_route: add `MIN_CLOCK_ROUTING_LAYER` + to `asap7` config

This is actually an integration of @maliberty changes to this pull request https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/564.

- asap7: adding `fastroute.tcl` to make larger design route

This is an opinionated `fastroute.tcl` file for asap7. I do run a bigger design as we have currently as designs for ASAP7. And without this fastroute script it hangs indefinitly on  [INFO GRT-0103] Extra Run for hard benchmark. This `fastroute.tcl` fixes it. But I first want to see what CI thinks of it :-).

joennlae committed:
floorplan: add `set_dont_use` before `repair_design`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
cts: move `set_dont_use` before `repair_clock_inverters`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
global_route: remove depreceated option `-report_violating_nets`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: remove unused `yosys` blackbox files

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: allow with `v28` working clock gating cells `ICG*`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: make `CTS_BUF_DISTANCE` editable via parent config

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
global_route: add `MIN_CLOCK_ROUTING_LAYER` + to `asap7` config

as discussed in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/564

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: make tapcell work with low and super low threshold

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: use correct gds, buffer, fille for corresponding vth

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: update `yosys` adder, clkgate, latch and dff for multi-vt

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: fix `.lef` file according to upstream repo

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: revert back to one buffer

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: update to `v28` `.lef` files

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: update to `v28` `.gds` files

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: update to `v28` adaption of `config.mk` file

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

maliberty commented on flow/platforms/asap7/fastroute.tcl:
```diff
@@ -0,0 +1,9 @@
+set_global_routing_layer_adjustment M3 0.3
+set_global_routing_layer_adjustment M4 0.3
+set_global_routing_layer_adjustment M5 0.3
+set_global_routing_layer_adjustment M6-$::env(MAX_ROUTING_LAYER) 0.2
```
How did you come up with these values?

joennlae committed:
asap7: remove `4x` gds + lef

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: remove `4x` option from `asap7`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
scripts: move `set_dont_use` into `load.tcl` and remove other refs

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

joennlae committed:
asap7: update `metadata` after `v28` lib upgrade

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>

ASAP7 aes

[INFO] Tightening rule synth__design__instance__area__stdcell from 2576 to 1918.
[INFO] Tightening rule placeopt__design__instance__area from 2875 to 2058.
[INFO] Tightening rule placeopt__design__instance__count__stdcell from 25423 to 17930.
[INFO] Updating failing rule cts__timing__setup__ws__pre_repair from -166.4 to -390.96.
[INFO] Updating failing rule cts__timing__setup__ws__post_repair from -166.4 to -390.96.
[INFO] Tightening rule cts__design__instance__count__setup_buffer from 1105 to 780.
[INFO] Tightening rule cts__design__instance__count__hold_buffer from 1105 to 780.
[INFO] Updating failing rule globalroute__timing__clock__slack from -35.7 to -68.66.
[INFO] Updating failing rule globalroute__timing__setup__ws from -35.7 to -68.67.
[INFO] Tightening rule detailedroute__route__wirelength from 116087 to 81427.
[INFO] Updating failing rule finish__timing__setup__ws from 0.0 to -49.41.
[INFO] Tightening rule finish__design__instance__area from 3035 to 2212.

ASAP7 ethmac

[INFO] Tightening rule synth__design__instance__area__stdcell from 8662 to 8647.
[INFO] Updating failing rule globalroute__timing__clock__slack from -124.58 to -178.17.
[INFO] Updating failing rule globalroute__timing__setup__ws from -124.58 to -178.18.
[INFO] Updating failing rule finish__timing__setup__ws from -191.62 to -326.37.
[INFO] Tightening rule finish__timing__drv__max_slew_limit from -1.55 to -1.16.

ASAP7 gcd

[INFO] Tightening rule synth__design__instance__area__stdcell from 44 to 40.
[INFO] Tightening rule placeopt__design__instance__area from 51 to 48.
[INFO] Tightening rule placeopt__design__instance__count__stdcell from 514 to 467.
[INFO] Tightening rule cts__timing__setup__ws from -150.09 to -145.94.
[INFO] Tightening rule cts__timing__setup__ws__pre_repair from -141.63 to -138.77.
[INFO] Tightening rule cts__timing__setup__ws__post_repair from -141.63 to -138.77.
[INFO] Tightening rule cts__design__instance__count__setup_buffer from 22 to 20.
[INFO] Tightening rule cts__design__instance__count__hold_buffer from 22 to 20.
[INFO] Tightening rule detailedroute__route__wirelength from 1378 to 1319.
[INFO] Tightening rule finish__design__instance__area from 54 to 51.

ASAP7 ibex

[INFO] Tightening rule placeopt__design__instance__area from 2729 to 2726.
[INFO] Tightening rule placeopt__design__instance__count__stdcell from 22552 to 22218.
[INFO] Tightening rule cts__design__instance__count__setup_buffer from 980 to 966.
[INFO] Tightening rule cts__design__instance__count__hold_buffer from 980 to 966.
[INFO] Tightening rule detailedroute__route__wirelength from 110102 to 106385.
[INFO] Tightening rule finish__design__instance__area from 2743 to 2739.
[INFO] Updating failing rule finish__timing__drv__max_slew_limit from -0.2 to -1.17.

ASAP7 jpeg

[INFO] Tightening rule synth__design__instance__area__stdcell from 7265 to 7234.
[INFO] Tightening rule placeopt__design__instance__area from 7580 to 7543.
[INFO] Tightening rule placeopt__design__instance__count__stdcell from 63235 to 62738.
[INFO] Tightening rule cts__timing__setup__ws from -273.77 to -271.57.
[INFO] Tightening rule cts__design__instance__count__setup_buffer from 2749 to 2728.
[INFO] Tightening rule cts__design__instance__count__hold_buffer from 2749 to 2728.
[INFO] Updating failing rule finish__timing__setup__ws from -255.53 to -332.2.
[INFO] Tightening rule finish__design__instance__area from 7608 to 7568.
[INFO] Updating failing rule finish__timing__drv__max_slew_limit from -0.2 to -0.48.

ASAP7 sha3

[INFO] Tightening rule finish__timing__setup__ws from -222.99 to -83.25.
[INFO] Tightening rule finish__timing__drv__max_slew_limit from -0.84 to -0.2.
[INFO] Tightening rule finish__timing__wns_percent_delay from -39.47 to -19.57.

ASAP7 uart

[INFO] Tightening rule synth__design__instance__area__stdcell from 136 to 80.
[INFO] Tightening rule placeopt__design__instance__area from 94 to 88.
[INFO] Tightening rule placeopt__design__instance__count__stdcell from 827 to 762.
[INFO] Tightening rule cts__design__instance__count__setup_buffer from 36 to 33.
[INFO] Tightening rule cts__design__instance__count__hold_buffer from 36 to 33.
[INFO] Tightening rule globalroute__timing__clock__slack from -56.86 to -56.68.
[INFO] Tightening rule globalroute__timing__setup__ws from -56.87 to -56.68.
[INFO] Tightening rule detailedroute__route__wirelength from 1901 to 1770.
[INFO] Tightening rule finish__design__instance__area from 126 to 95.

maliberty commented on flow/scripts/global_route.tcl:
```diff
@@ -10,7 +10,11 @@ if {[info exist env(FASTROUTE_TCL)]} {
+    set_routing_layers -signal $env(MIN_ROUTING_LAYER)-$env(MAX_ROUTING_LAYER) -clock $env(MIN_CLOCK_ROUTING_LAYER)-$env(MAX_ROUTING_LAYER)
+  } else {
+    set_routing_layers -signal $env(MIN_ROUTING_LAYER)-$env(MAX_ROUTING_LAYER)
+  }
```
Simplify:
```
 set_routing_layers -signal $env(MIN_ROUTING_LAYER)-$env(MAX_ROUTING_LAYER)
 if {[info exist env(MIN_CLOCK_ROUTING_LAYER)]} {
    set_routing_layers  -clock $env(MIN_CLOCK_ROUTING_LAYER)-$env(MAX_ROUTING_LAYER)
  } 
```

maliberty commented on flow/platforms/asap7/config.mk:
```diff
@@ -99,18 +96,20 @@ export PLACE_DENSITY ?= 0.60
 
 # TritonCTS options
-export CTS_BUF_CELL            ?= BUFx4_ASAP7_75t_R
+export CTS_BUF_CELL            = BUFx4_ASAP7_75t_R
```
Why remove the ?

joennlae replied:
I wanted to be able to change the buffer cell via my design config. I can also remove it but this would make that possible.

joennlae replied:
Ah, wait yeah, you are right let me add that again :-)

maliberty commented on flow/platforms/asap7/config.mk:
```diff
@@ -99,18 +96,20 @@ export PLACE_DENSITY ?= 0.60
 # Fill cells used in fill cell insertion
 export FILL_CELLS              = "FILLERxp5_ASAP7_75t_R"
 
+export TAP_CELL_NAME					 = TAPCELL_ASAP7_75t_R
```
avoid tabs

maliberty commented on flow/platforms/asap7/config.mk:
```diff
@@ -120,40 +119,107 @@ export KLAYOUT_TECH_FILE       = $(PLATFORM_DIR)/KLayout/asap7.lyt
 
 # XS - defining function for using LVT
-ifdef ($(ASAP7_USELVT))
+ifeq ($(ASAP7_USELVT), 1)
```
Why?

joennlae replied:
It did not work for me. And we use `ifeq` for the rest of the `config.mk` file.

maliberty commented on flow/platforms/asap7/config.mk:
```diff
@@ -120,40 +119,107 @@ export KLAYOUT_TECH_FILE       = $(PLATFORM_DIR)/KLayout/asap7.lyt
+
+	 export FILL_CELLS              = "FILLERxp5_ASAP7_75t_L"
+
+	 export TAP_CELL_NAME					  = TAPCELL_ASAP7_75t_L
```
avoid tabs to prevent alignment issues

joennlae committed:
asap7: apply changes from `review`

Signed-off-by: Jannis Schönleber <joennlae@gmail.com>
